third-party/fb-mysql/8.0.20/man/zlib_decompress.1
'\" t
.\"     Title: \fBzlib_decompress\fR
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 03/06/2020
.\"    Manual: MySQL Database System
.\"    Source: MySQL 8.0
.\"  Language: English
.\"
.TH "\FBZLIB_DECOMPRESS\FR" "1" "03/06/2020" "MySQL 8\&.0" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
zlib_decompress \- decompress mysqlpump ZLIB\-compressed output
.SH "SYNOPSIS"
.HP \w'\fBzlib_decompress\ \fR\fB\fIinput_file\fR\fR\fB\ \fR\fB\fIoutput_file\fR\fR\ 'u
\fBzlib_decompress \fR\fB\fIinput_file\fR\fR\fB \fR\fB\fIoutput_file\fR\fR
.SH "DESCRIPTION"
.PP
The
\fBzlib_decompress\fR
utility decompresses
\fBmysqlpump\fR
output that was created using ZLIB compression\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
If MySQL was configured with the
\fB\-DWITH_ZLIB=system\fR
option,
\fBzlib_decompress\fR
is not built\&. In this case, the system
\fBopenssl zlib\fR
command can be used instead\&.
.sp .5v
.RE
.PP
Invoke
\fBzlib_decompress\fR
like this:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBzlib_decompress \fR\fB\fIinput_file\fR\fR\fB \fR\fB\fIoutput_file\fR\fR
.fi
.if n \{\
.RE
.\}
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBmysqlpump \-\-compress\-output=ZLIB > dump\&.zlib\fR
shell> \fBzlib_decompress dump\&.zlib dump\&.txt\fR
.fi
.if n \{\
.RE
.\}
.PP
To see a help message, invoke
\fBzlib_decompress\fR
with no arguments\&.
.PP
To decompress
\fBmysqlpump\fR
LZ4\-compressed output, use
\fBlz4_decompress\fR\&. See
\fBlz4_decompress\fR(1)\&.
.SH "COPYRIGHT"
.br
.PP
Copyright \(co 1997, 2020, Oracle and/or its affiliates. All rights reserved.
.PP
This documentation is free software; you can redistribute it and/or modify it only under the terms of the GNU General Public License as published by the Free Software Foundation; version 2 of the License.
.PP
This documentation is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along with the program; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA or see http://www.gnu.org/licenses/.
.sp
.SH "SEE ALSO"
For more information, please refer to the MySQL Reference Manual,
which may already be installed locally and which is also available
online at http://dev.mysql.com/doc/.
.SH AUTHOR
Oracle Corporation (http://dev.mysql.com/).


NEWS
"Rogers" 10-Sep-2015
- Update to jemalloc 4.0.0.
- Convert more architecture dependent code to VASM.
- Large bout of fixes for compiling hhvm under MSVC.
- Have hh_emitter support more of HHBC.
- Remove MInstrState from CGetM and IssetM's final operations.
- Refactor [un]serialize functions.
- Use profiling counters to decide whether to apply LICM.
- Move inlining to translateRegion.
- Implement intdiv function.
- Remove joomla and joomlacms from hhvm test suite.
- [hack] Add a --stats option to the client.
- [hack] GC the shared memory area more aggressively.
- [hack] Rework the representation of strings.
- Fix bug #70014 - use RAND_bytes instead of deprecated RAND_pseudo_bytes
- Fix bug #69545 - avoid overflow when reading list and improve fix for bug #69545
- Fix #66387: Stack overflow with imagefilltoborder
- Fix #70064: imagescale(..., IMG_BICUBIC) leaks memory and Fix #69024: imagescale segfault with palette based image

"Quake" 27-Aug-2015
- ??

"Park" 13-Aug-2015
- ???

"Osborne" 30-Jul-2015
- Remove global LDAP requirement
- Don't require libXSLT globally
- Remove the global MCrypt requirement
- Remove the global search for LibUODBC
- Remove the global check for libpam
- use folly::writeFull wrapper for uploads
- Fix vararg handling for array_uintersect and others
- fix error message for abstract type constants
- Fix variadics in hhvm_get_frame_args
- Use poll() instead of select() in XDebugServer::createSocket.
- [hack] Error on overriding concrete type constant with an abstract one
- [hack] Grow shapes more correctly
- Array pointer modification functions (next, current, each etc) can be used on ArrayObject
- Eliminate ext_mbstring's dependence on ext_process
- Eliminate ext_mail's dependence on ext_process
- Use utf8 string literals in zend-html
- Don't allow abstract memoized methods
- [hack] Properly handle 'this' type for methods
- Fix lexing of "${fn()}"
- Fix lexing of #<?php
- [hack] Fix infinite loop in syntax error when closing a namespace
- Fix mispelling of COOKIE.
- Deal with self, static and parent constructors

"Nash" 16-Jul-2015
- Make ArrayData ref-count start at 1
- Remove the NOUSER define when compiling with MSVC, Cygwin, or MinGW
- Fix typos of `__MSC_VER` to be the correct value, `_MSC_VER`
- Shapes::removeKey(...) - remember the names of unset fields
- Shapes::toArray() - add .hhi
- Shapes::removeKey()
- Shapes::keyExists()
- Shapes::idx helper function for accessing optional fields
- Don't assume a scalar Map/Set initializer is an integer if its not a string
- Update mysqlclient library to 5.6.23 everywhere
- Port fix for php bug #48147
- Fixed #5475: ResourceBundle::getLocales('') don't return list of locales
- [hhvm][oss][frameworks] Add Smarty to Framework test runner
- Add in a facility to pool curl handles between requests.
- Fix xdebug trying to write to a nullptr std::FILE*.
- Fix bug when appending unaligned StringSlice
- Remove XDebug chrome support
- Incorrect return value from eval, Closure generated in first eval pass is returned in the second eval pass
- Stop throwing ints in xdebug
- CVE-2015-4663: Support verifying certificates on ssl/tls streams, enable by default
- Allow inlining immutable methods in non-repo mode
- Reusable TC
- Release some old hack design documentation
- Don't step into builtin functions in the debugger
- Support collections in array_map.hhas
- [hhvm] Make sure iteration of class_iterator actually stops
- [hhvm] Open source idx
- Add .hhas implementation of array_reduce
- Change memory layout for hni

"McQueen" 2-Jul-2015
- Add basic stream_socket_enable_crypto() client support
- [hack] Check types for class vars in decl mode as well
- [hhvm] Fix xdebug with multiple commands (and hopefully with PhpStorm)
- [hack][hhvm] Standard library for shapes - runtime support
- Fix date_create_immutable error state
- [hack] Include debug symbols in OSS build
- Update Hack Tools README to point to new OPAM Quick Install URL
- [hack] add color mode to Tty
- Update CMake/FindTBB.cmake for current Windows support.
- [hhvm][mac] Fix ext_ipc on Mac
- [hhvm][easy] Add README.md to php-octane
- [zoncolan] Missing cases to bring BW analysis at par with FW
- [zoncolan] Added "support" for reflection
- [zoncolan] Added support for XHP
- [hack][shapes] Anonymous shapes
- [hack][shapes] Structural subtyping for shapes proposal
- Fix some bugs in ext_servicerouter.cpp / ext_gatehouse.cpp
- [zoncolan] Added analysis of toplevel functions
- [zonconal] Adding shapes
- [hack] Infer types for constants in decl mode as well
- [hhvm] Enable JIT by default on Mac
- [hack] Format: Improve handling of comments in switch statements
- Make ObjectData ref-counts start at 1
- APC collection optimizations
- [Hack] Fix JSON escaping
- [hack] Make Typing_instantiate only work on decl ty
- use ...$args in invariant_violation()
- [hack] Make decl mode equivalent to partial + // UNSAFE
- [hhvm] Don't use libbfd in OSS build by default
- [hhvm][mac] Fix async mysql time calculation on Mac clang

"Lukather" 18-Jun-2015
- Check that the right operand of instanceof is not an enum
- Allow inlining static builtin methods
- Fix hhbbc not emitting RATs for public static properties.
- Fix overflow in array_slice
- Better error message for double quotes used in shape field names
- infer return type of floor/ceil
- Do not translate functions from units marked as interpret-only.
- Eliminate ldimm saveflags operand
- Allow folding the ideomatic hack callable wrappers
- Optimize NativeImpl
- Modify ReleaseVVOrExit to deal with VarEnvs
- Allow inlined 86ctors
- Fix folly build with Clang
- Re-fix region-prune-arcs assertion
- [mac] Fix crash compiling crc32 with Mac clang
- [mac] Use -force_load, not -all_load
- Don't constrain key or value for SetWithRefElem
- Remove IRBuilder::prepareForNextHHBC---just use exceptionStackBoundary
- Use HHBBC types to avoid generic IncRefs in Await translations
- Don't check for failed waithandles until after it wasn't finished

third-party/fb-mysql/8.0.20/utilities/zlib_decompress.cc

/*
   Copyright (c) 2014, 2019, Oracle and/or its affiliates. All rights reserved.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 2.0,
   as published by the Free Software Foundation.

   This program is also distributed with certain software (including
   but not limited to OpenSSL) that is licensed under separate terms,
   as designated in a particular file or component or in included license
   documentation.  The authors of MySQL hereby grant you an additional
   permission to link the program and your derivative works with the
   separately licensed software that they have included with MySQL.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License, version 2.0, for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <zlib.h>

#include "my_sys.h"
#include "print_version.h"
#include "welcome_copyright_notice.h"

static void usage() {
  print_version();
  puts(ORACLE_WELCOME_COPYRIGHT_NOTICE("2015"));
  puts(
      "Decompress data compressed by mysqlpump using zlib compression "
      "algorithm by reading from input file and writing uncompressed "
      "data to output file");
  printf("Usage: %s input_file output_file\n", "zlib_decompress");
}

static const int INPUT_BUFFER_SIZE = 1024 * 1024;
static const int OUTPUT_BUFFER_SIZE = 1024 * 1024;

int main(int argc, char **argv) {
  MY_INIT(argv[0]);
  if (argc != 3) {
    usage();
    exit(1);
  }
  FILE *input_file = fopen(argv[1], "rb");
  FILE *output_file = fopen(argv[2], "wb");
  if (input_file == nullptr) {
    fprintf(stderr,
            "zlib_decompress: [Error] Cannot open input file for reading.\n");
    exit(1);
  }
  if (output_file == nullptr) {
    fprintf(stderr, "zlib_decompress: [Error] Cannot create output file.\n");
    exit(1);
  }
  z_stream decompression_context;
  memset(&decompression_context, 0, sizeof(decompression_context));
  inflateInit(&decompression_context);
  char *input_buffer = new char[INPUT_BUFFER_SIZE];
  char *output_buffer = new char[OUTPUT_BUFFER_SIZE];
  bool is_eof = false;

  while (!is_eof) {
    int bytes_read = fread(input_buffer, 1, INPUT_BUFFER_SIZE, input_file);

    if (bytes_read < INPUT_BUFFER_SIZE) {
      is_eof = feof(input_file);
      if (!is_eof) {
        fprintf(stderr,
                "zlib_decompress: [Error] Encountered problem during reading "
                "input.\n");
        exit(1);
      }
    }

    decompression_context.avail_in = bytes_read;
    decompression_context.next_in = (Bytef *)input_buffer;

    do {
      decompression_context.avail_out = OUTPUT_BUFFER_SIZE;
      decompression_context.next_out = (Bytef *)output_buffer;
      int res = inflate(&decompression_context, Z_NO_FLUSH);
      if (res == Z_STREAM_ERROR) {
        fprintf(stderr,
                "zlib_decompress: [Error] Encountered problem during "
                "decompression.\n");
        exit(1);
      }

      fwrite(output_buffer, 1,
             OUTPUT_BUFFER_SIZE - decompression_context.avail_out, output_file);
    } while (decompression_context.avail_out == 0);
  }

  inflateEnd(&decompression_context);
  delete[] input_buffer;
  delete[] output_buffer;
  fclose(input_file);
  fclose(output_file);
  my_end(0);
  exit(0);
}
